pkg/lib/oauth/token_encoding.go

```diff
@@ -3,19 +3,102 @@ package oauth
 import (
 	"errors"
 	"fmt"
+	"net/url"
 	"strings"
+	"time"
+
+	"github.com/lestrrat-go/jwx/jwa"
+	"github.com/lestrrat-go/jwx/jwt"
+
+	"github.com/authgear/authgear-server/pkg/lib/config"
+	"github.com/authgear/authgear-server/pkg/util/clock"
+	"github.com/authgear/authgear-server/pkg/util/jwkutil"
+	"github.com/authgear/authgear-server/pkg/util/jwtutil"
 )
 
-func EncodeAccessToken(token string) string {
-	return token
+type UserClaimsProvider interface {
+	LoadUserClaims(userID string) (jwt.Token, error)
 }
 
-func EncodeRefreshToken(token string, grantID string) string {
-	return fmt.Sprintf("%s.%s", grantID, token)
+type BaseURLProvider interface {
+	BaseURL() *url.URL
+}
+
+type AccessTokenEncoding struct {
+	Secrets    *config.OAuthKeyMaterials
+	Clock      clock.Clock
+	UserClaims UserClaimsProvider
+	BaseURL    BaseURLProvider
+}
+
+func (e *AccessTokenEncoding) EncodeAccessToken(client *config.OAuthClientConfig, grant *AccessGrant, userID string, token string) (string, error) {
+	if !client.UseJWTAccessToken {
+		return token, nil
+	}
+
+	claims, err := e.UserClaims.LoadUserClaims(userID)
+	if err != nil {
+		return "", err
+	}
+
+	// iss & sub claims are in user claims.
+	_ = claims.Set(jwt.AudienceKey, e.BaseURL.BaseURL().String())
```

Comment:
In ID token, we put client ID as `aud`. Should we do the same?

Reply:
The audience of access token is the resource server (i.e. ourselves). Clients are not supposed to interpret access tokens, therefore it's inappropriate to include them in audiences.

Reply:
But I think one of the motivation to have introduce JWT access token is to allow the developer's backend server to see the claims. So are they also audience?

Reply:
In that case, the backend server is a resource server, so it should be included in audiences. Referring this scenario:
- Developer has authorization server (i.e. authgear)
- Developer has a backend API server (i.e. resource server)
- Developer has a iOS app (i.e. OAuth client)
- Developer has an Android app (i.e. OAuth client)

It can be observed that resource servers are not related to OAuth clients. Instead, according to the [draft spec](https://tools.ietf.org/html/draft-ietf-oauth-access-token-jwt-10#section-3), the authorization server should know a list of resource server indicators. However, I'm hesitant to add such a list to config at the moment, since it's unclear whether it would be useful.